keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,210 @@
+- Platform specific options are mixed together. For example, the IPTables rule sync fields are used by the Windows HNS backend for the same purpose.
+- Again, the IPTables rule sync options are used for the Linux legacy user mode proxy, but not for the IPVS mode (where a set of identical options exist, despite the fact, that it too uses some other fields, designed for IPTables).
+
+Clearly, this made the configuration both hard to use and to maintain. Therefore, a plan to restructure and stabilize the config format is needed.
+
+### Goals
+
+- To cleanup the existing config format.
+- To provide config structure, that is easier for users to understand and use.
+- To distinguish between instance local and shared settings.
+- To allow for the persistence of settings for different platforms (such as Linux and Windows) in a manner that reduces confusion and the possibility of an error.
+- To allow for easier introduction of new proxy backends.
+- To provide users with flexibility, especially with regards to the config source.
+
+### Non-Goals
+
+- To change or implement additional features in kube-proxy.
+- To deal with graduation of any other component of kube-proxy, other than its configuration.
+- To remove most or even all of the command line flags, that have corresponding component config options.
+
+## Proposal
+
+The idea is to conduct the process of graduation to beta in small steps in the span of at least one Kubernetes release cycle. This will be done by creating one or more alpha versions of the config with the last alpha version being renamed to v1beta1 after the community is happy with it.
+Each of the sub-sections below can result in a separate alpha version release, although it will be better for users to have no more than a couple of alpha versions past v1alpha1.
+The proposed process is similar to the already successfully used one for kubeadm.
+
+### Re-encapsulate mode specific options
+
+The current state of the config has proven that:
+- Some options are deemed as mode specific, but are in fact shared between all modes.
+- Some options are placed directly into KubeProxyConfiguration, but are in fact mode specific ones.
+- There are options that are shared between some (but not all) modes. Specific features of the underlying implementation are common and this happens only within the boundaries of the platform (iptables and ipvs modes for example).
+- Although legacy Linux and Windows user mode proxies are separate code bases, they have a common set of options.
+
+With that in mind, the following measures are proposed:
+- Mode specific structs are consolidated to not use fields from other mode specific structs.
+- Introduce a single combined legacy user mode proxy struct for both Linux and Windows backends.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
```

Comment:
I'm not convinced of this grouping - it seems substantially more complex and error prone for users, compared duplicating platform-specific fields into their respective modes. Is there an additional reason that I'm missing?

Reply:
This tries to solve the case where we have a mixed cluster between Linux & Windows nodes. In the normal case we would like to have one of the kernel modes for Linux enabled (IPTables or IPVS) and the Windows kernel mode support (HNS).
In that case we are left with one of these options:
- Have a couple of different configs (one for Linux and one for Windows nodes). They can be housed in different config maps as well.
  - *Pros:* Very simple and easy to understand.
  - *Cons:* Platform independent fields are very easy deviate between the config, thus causing errors.
- Use a single config, but initialize settings for corresponding modes.
  - *Pros:* Platform independent fields are now shared.
  - *Cons:* Deceptively simple (in reality it isn't). How do you tell which mode to use on which OS?
- Use a single config with the proposed platform splitting.
  - *Pros:* Platform independent fields are again shared and there is a clear distinction between what mode applies to what platform.
  - *Cons:* Somewhat complex looking indeed.

Hence, I see the last option as probably the best to choose here. However, I am not sold only to this option. If we consider, that the "mixed platform" cluster is a rare enough use case, and not worth the hassle of introducing a platform split, then I am happy with not doing anything at all.

Reply:
we could multi-yaml docs for the separate OS:
```
KubeProxySharedConfiguration
---
KubeProxySharedConfigurationWindows
---
KubeProxySharedConfigurationLinux
---
```

and write all these in a ConfigMap for all consumers.

```
KubeProxyInstanceConfiguration
---
KubeProxyInstanceConfigurationXXXXXX
---
```

these we can have "on disk" and be instance specific, where KubeProxyInstanceConfiguration contains the OS agnostic config and KubeProxyInstanceConfigurationXXXXXX is the OS specific config for this instance.

Reply:
The mixed platform case is going to be common enough to care about (think about all the enterprises running Windows stuff).

I recommend that config for a pool of nodes one one platform just be in a separate wrapper (ConfigMap, file, whatever) vs. the config for a pool on another platform. This gives you a small, targeted config that is much easier to read when debugging, because you've only included relevant fields.

Reply:
If we have a `ProxyMode`, it's going to be placed in the platform specific options. We may even omit the `ProxyMode` option and instead force the user to provide only one mode sub-struct filled on a per-platform basis (error out on more than one). That's an implementation detail to me.

On a recap, having platform-independent modes is not going to be a good idea (especially if they are the legacy user modes). I don't think, that new platform independent modes will be added to kube-proxy. Hence, it's safe to say, that we can treat all modes as platform specific and duplicate them if necessary.

I don't think, that kube-proxy should have any understanding of where the config came from. It should simply receive paths to YAML files on its command line and read them. This way we can have different pools of instances. Some can take their config from a config map volume source, named "KubeProxyConfigA", others can take it from "KubeProxyConfigB". A third set can use an init container to generate it, etc.

Reply:
> We may even omit the ProxyMode option and instead force the user to provide only one mode sub-struct filled on a per-platform basis (error out on more than one).

This is the experience I'd lean towards, because it encourages configs to be small and specific.

> I don't think, that new platform independent modes will be added to kube-proxy. Hence, it's safe to say, that we can treat all modes as platform specific and duplicate them if necessary.

Sounds good. 

> I don't think, that kube-proxy should have any understanding of where the config came from. It should simply receive paths to YAML files on its command line and read them. This way we can have different pools of instances. Some can take their config from a config map volume source, named "KubeProxyConfigA", others can take it from "KubeProxyConfigB". A third set can use an init container to generate it, etc.

Cool, that's what I was thinking too.

Reply:
I still think this is overly-granular for users, compared to the config normalization that it provides.